src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
EnumAllVirtualSlots won't see them. We would still need magic special casing (GetClosestMetadataType again?) when building vtables. And the logic around GetVirtualMethodSlot will be off because the owning type is a weird type.

What's the argument against these methods showing up in GetAllMethods? Can you use the same argument when we talk about RuntimeInterfaces on array types? What makes it different?

The way I look at it - a type system consumer shouldn't have to care about magicness of the array types. If it walks like a duck and quacks like a duck, it is a duck (if it implements the interface and has interface implementation methods, the type has the methods).